OGM-249 First draft of CouchBase backend

ewa01s opened the pull request:
I Implemented CRUD for tuples - first phase of backend for CouchBase, is it possible to assign task ([OGM-249](https://hibernate.atlassian.net/browse/OGM-249)) to me ([ewastawicka](https://hibernate.atlassian.net/secure/ViewProfile.jspa?name=ewastawicka))?

gunnarmorling commented:
@ewa01s, many thanks for following up on this! If you haven't done so, could you sign the JBoss [contributor license agreement](https://cla.jboss.org/index.seam)? We need that so we can integrate your work.

gunnarmorling commented:
ok to test.

gunnarmorling commented:
@ewa01s, you are assigned to the right group in JIRA now, the issue is assigned to you :)

gunnarmorling commented:
Going to check it out in the afternoon.

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/dialect/backend/json/impl/EntityDocument.java:
```diff
@@ -0,0 +1,234 @@
+ */
+@JsonSerialize(include = Inclusion.NON_NULL)
+@JsonTypeName(EntityDocument.TYPE_NAME)
+public class EntityDocument {
```
This looks very similar to the implementation in the _couchdb_ module. Maybe we could extract a common implementation and use it for both?

ewa01s commented:
> This looks very similar to the implementation in the couchdb module. Maybe we could extract a common implementation and use it for both?

I tried, the problem for me was how to keep revision field in all inheritied couchdb classes and not in couchbase classes, is there a better solution than [b2c2910](https://github.com/ewa01s/hibernate-ogm/commit/b2c2910d1772a05a96bb666b81bcf0bb0d5c848c)?

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/dialect/backend/json/impl/CouchBaseDocument.java:
```diff
@@ -0,0 +1,108 @@
+/**
+ * @author Stawicka Ewa
+ */
+public class CouchBaseDocument extends AbstractDocument {
```
So what is the difference to a CouchDB document?

ewa01s replied:
CouchBase doesn't have support for versioning documents, so I think it doean't make sense to keep revision field in the instances of CouchBase documents.

ewa01s replied:
That problem doesn't exist anymore since I applied another solution for creating object for CouchBase Java Api.

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/dialect/type/impl/CouchBaseTypeContributor.java:
```diff
@@ -0,0 +1 @@

	}

}
```
Is that one even needed?

ewa01s replied:
No ;)

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/impl/CouchBaseDatastoreProvider.java:
```diff
@@ -0,0 +1,103 @@
+
+	@Override
+	public boolean allowsTransactionEmulation() {
+		// TODO Auto-generated method stub
```
This should be true; That we it can be used with `RESOURCE_LOCAL` transaction type instead of requiring a JTA implementation.

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/logging/impl/LoggerFactory.java:
```diff
@@ -0,0 +1,17 @@
+ */
+public class LoggerFactory {
+	public static Log getLogger() {
+		return Logger.getMessageLogger( Log.class, "CouchBase" );
```
Better pass the caller class name as a category, that's what we do for (most of) the other backends. E.g. check out the logger factory in core.

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/util/impl/DatabaseIdentifier.java:
```diff
@@ -0,0 +1,68 @@
+	private final String password;
+	private final URI serverUri;
+	private final URI databaseUri;
+	public DatabaseIdentifier(String host, int port, String databaseName, String password)
```
Seems like its copied from the CouchDB backend? It'd be nice if we could unify it.

gunnarmorling commented on couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/logging/impl/Log.java:
```diff
@@ -0,0 +1,36 @@
+@MessageLogger(projectCode = "OGM")
+public interface Log extends org.hibernate.ogm.util.impl.Log {
+	@LogMessage(level = INFO)
+	@Message(id = 1901, value = "Connecting to CouchBase at %s")
```
You can start with 1701; Currently 1601 - 1700 is the upermost interval in use (by Cassandra).

gunnarmorling commented on couchbase/src/test/java/org/hibernate/ogm/datastore/couchbase/utils/CouchBaseTestHelper.java:
```diff
@@ -0,0 +1,97 @@
+	@Override
+	public long getNumberOfEntities(SessionFactory sessionFactory) {
+		// TODO Auto-generated method stub
+		return 0;
```
That should be implemented to make some of the tests work.

gunnarmorling commented:
Nice work, @ewa01s :) Many thanks! Some general thoughts/questions:

* It would be great if we could avoid duplications between CouchDB and CouchBase as far as possible
* Each class should have a class-level comment describing its purpose, important design decisions and alike
* How does CouchBase deal with id sequences? Is there such thing at all? Or should one preferably work with UUIDs for instance?
* Could you exclude all those tests of the backend TCK not yet passing? You can e.g. check out the Cassandra POM file which explicitely includes only a single test atm. (I guess some more will be passing in your case). So we have a build which actually succeeds, we then can include more and more tests of the suite as more features (e.g. associations) get implemented.

Sanne commented:
I just noticed there's a new commit: @ewa01s sorry we didn't notice as github won't notify anyone if you add a commit on an existing pull request.
What's the update? I see ci.hibernate.org are quite happy, did you address all of @gunnarmorling 's suggestions?

ewa01s commented:
@Sanne thanks for pointing me! Next time when I add a commit I will add a comment as well.

- [x] I changed approach of connecting, now CouchBaseDialect works on CouchBase Java API models (JsonDocument, JsonObject) so we don't need Jackson, and APIs from couchdb module anymore :)
- [x] I excluded tests which failure

>did you address all of @gunnarmorling 's suggestions?

Not yet, I'm still working on:
- [ ] change logger factory
- [ ] move DatabaseIdentifier and Identifier to common module
- [ ] getNumberOfEntities
- [ ] better documentations of classes

>How does CouchBase deal with id sequences? Is there such thing at all? Or should one preferably work with UUIDs for instance?

Yes, it has [Atomic operations](http://docs.couchbase.com/developer/java-2.1/documents-atomic.html), we can make use of that.

I would also like to develop, in this pr, assosiations, we should have then quite powerful module.

gunnarmorling commented:
@ewa01s Glad to hear that you are making progress. Wrt. to associations I'd say let's focus on everything else in this PR and then have another one for association handling. No need to do everything at once, and the larger a single PR gets, the tougher it becomes to review it :)

That being said, I'd expect quite a few tests to fail due to the missing assoc support. Can you exclude those in the POM temporarily similar to how its done for Cassandra? That way a `mvn clean install` should pass without failure. You can then re-enable those tests with the assoc handling PR.

ewa01s commented:
@gunnarmorling, ok, associations will be implemented later. I've already excluded test which failure, currently 20 of integration tests work for build of <i>couchbase</i>

ewa01s commented:
I implemented API for id sequences,
I changed LoggerFactory,
I moved method createId from CouchDBDatastore to the Identifier. I moved Identifier to core module, and then replaced fromColumnValues, escapeCharsValuesUsedAsColumnValuesSeparator with methods from StringHelper, any issues not to do that?

Still on the list:
- [ ] Implement counting number of entities
- [ ] Ability do create database (i.e bucket) if requested

DavideD commented:
@ewa01s Because of a small change in the DocumentStoreConfiguration the current code does not compile with the latest master.  Could You fix it? You only need to change the qya you read the host and the port.

It should be something like this:
```
String host = configuration.getHosts().getFirst().getHost();
Integer port = configuration.getHosts().getFirst().getPort();
```

Thanks

ewa01s commented:
@DavideD thanks, I corrected it.

gunnarmorling commented:
Hey @ewa01s, I don't expect you have any plans to continue work on this dialect?

I'm going to close this pull request for now in order to clean up a bit. Should you want to work again on this, that'd be great of course. I'd suggest to get in touch over the mailing list or the chat on IRC so we can discuss next steps. Thanks!
